### PR TITLE
AK+LibDSP+Piano: Move Track to LibDSP (and all the yaks that are attached to that)

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
something something #6528 

This PR is now mergeable, as I have decided the lower two are not necessary features and might be added later. The purpose of this PR is just to get the Track move over with so that we can make smaller incremental progress. Therefore, the last :yaktangle: commit contains all the things to make Piano barely functional with a moved Track.

- [x] Audio plays
- [x] Note setting works
- [x] Audio player loop doesn't stop for indiscernible reasons

### CODE REVIEW

File: AK/FixedArray.h [hunk 1]
```diff
@@ -160,6 +160,13 @@ class FixedArray {
         ::swap(m_elements, other.m_elements);
     }
 
+    void fill_with(T value)
+    {
+        // FIXME: Let's hope the compiler optimizes this, otherwise we could do some trickery for trivial classes.
```
Comment (sin-ack) [inline AK/FixedArray.h:165]: The compiler should definitely be smart enough to optimize this. I'd recommend looking at the assembly before inserting FIXMEs. Ideally FIXMEs should be definite directives to fix an already-known issue.

File: Userland/Libraries/LibAudio/ConnectionFromClient.cpp [hunk 2]
```diff
@@ -62,6 +62,9 @@ ErrorOr<void> ConnectionFromClient::async_enqueue(FixedArray<Sample>&& samples)
     update_good_sleep_time();
     m_user_queue->append(move(samples));
     // Wake the background thread to make sure it starts enqueuing audio.
+    // FIXME: This can lead to a deadlock if the thread has gone to sleep but didn't set the flag yet.
+    //        If our caller depends on samples being played before it calls async_enqueue again (as is usually the case),
+    //        it will never call into this function again and we don't get a chance to actually wake the enqueuer thread.
```
Comment (sin-ack) [inline Userland/Libraries/LibAudio/ConnectionFromClient.cpp:67]: Your explanation in the commit message raises the question of why we don't set an event queue wake regardless of whether the thread is awake or not. Is this intended as an optimization?

Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibAudio/ConnectionFromClient.cpp:67]: Yes. If we send a wake every time, those just queue up as the event loop doesn't run as long as the enqueuer thread isn't finished playing audio. Therefore, as soon as we get a single break of audio, there would be a huge spike in useless calls into the custom event handler from all the queued event.

**Comment by kleinesfilmroellchen** [comment]
Converting to draft until I implement a better solution to the queue waking as discussed on Discord.

File: Userland/Libraries/LibCore/EventLoop.cpp [hunk 3]
```diff
@@ -6,6 +6,7 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include "AK/RefPtr.h"
```
Comment (sin-ack) [inline Userland/Libraries/LibCore/EventLoop.cpp:9]: clang powaaaaaaa

File: Userland/Libraries/LibCore/EventLoop.cpp [hunk 4]
```diff
@@ -501,6 +502,22 @@ void EventLoop::post_event(Object& receiver, NonnullOwnPtr<Event>&& event, Shoul
...
+void EventLoop::wake_once(Object& receiver, int custom_event_type)
+{
+    Threading::MutexLocker lock(m_private->lock);
+    dbgln_if(EVENTLOOP_DEBUG, "Core::EventLoop::wake_once: event type {}", custom_event_type);
+    auto identical_events = m_queued_events.find_if([&](auto& queued_event) {
+        if (queued_event.receiver.is_null())
+            return false;
+        auto const& event = queued_event.event;
+        return event->type() == Event::Type::Custom && static_cast<CustomEvent const*>(event.ptr())->custom_type() == custom_event_type
+            && queued_event.receiver.ptr() == &receiver;
```
Comment (sin-ack) [inline Userland/Libraries/LibCore/EventLoop.cpp:514]:
```suggestion
        auto is_receiver_identical = queued_event.receiver.ptr() == &receiver;
        auto event_id_matches = event->type() == Event::Type::Custom
            && static_cast<CustomEvent const*>(event.ptr())->custom_type() == custom_event_type;
        return is_receiver_identical && event_id_matches;
```
Less complex lines = more better.

File: Userland/Libraries/LibCore/EventLoop.cpp [hunk 5]
```diff
@@ -501,6 +502,22 @@ void EventLoop::post_event(Object& receiver, NonnullOwnPtr<Event>&& event, Shoul
...
+    });
+    // Event is not in the queue yet, so we want to wake. Otherwise, do nothing.
```
Comment (sin-ack) [inline Userland/Libraries/LibCore/EventLoop.cpp:516]:
```suggestion
    // Event is not in the queue yet, so we want to wake.
```

File: Userland/Libraries/LibCore/EventLoop.cpp (hunk 3 above)
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibCore/EventLoop.cpp:9]: jesus christ what is this program thinking

**Review by sin-ack — APPROVED** [review]

**Comment by linusg** [comment]
Conflicts :catdogalert:

**Comment by kleinesfilmroellchen** [comment]
@linusg :catdogalert: will fix later

File: AK/FixedArray.h [hunk 6]
```diff
@@ -160,6 +160,12 @@ class FixedArray {
         ::swap(m_elements, other.m_elements);
     }
 
+    void fill_with(T value)
```
Comment (linusg) [inline AK/FixedArray.h:163]: can we make this a const reference to avoid the one pointless copy?

Comment (kleinesfilmroellchen) [reply AK/FixedArray.h:163]: godbolt says this is optimized to vectorized memcopy, but sure

**Pull request merged**